Plugins/NativeCamera/Source/Android/NativeCameraImpl.cpp

```diff
@@ -309,129 +256,248 @@ namespace Babylon::Plugins
 
                 // Start capturing continuously
-                GET_CAMERA_FUNCTION(ACameraCaptureSession_setRepeatingRequest)(m_textureSession,
+                GET_CAMERA_FUNCTION(ACameraCaptureSession_setRepeatingRequest)(m_implData->textureSession,
                                                                                &captureCallbacks,
-                                                                               1, &m_request,
+                                                                               1, &m_implData->request,
                                                                                nullptr);
             }
 
-            if (eglMakeCurrent(m_display, 0/*surface*/, 0/*surface*/, currentContext) == EGL_FALSE)
+            if (eglMakeCurrent(m_implData->display, 0/*surface*/, 0/*surface*/, currentContext) == EGL_FALSE)
             {
                 throw std::runtime_error{"Unable to restore GL context for camera texture init."};
             }
 
-            return m_cameraDimensions;
+            return m_implData->cameraDimensions;
         });
     }
 
+    std::vector<std::shared_ptr<CameraDevice>> Camera::Impl::GetCameraDevices()
+    {
+        std::vector<std::shared_ptr<CameraDevice>> cameraDevices{};
+
+        // Get the list of available cameras
+        API24::ACameraIdList *cameraIds = nullptr;
+        GET_CAMERA_FUNCTION(ACameraManager_getCameraIdList)(m_implData->cameraManager, &cameraIds);
+
+        for (int i = 0; i < cameraIds->numCameras; ++i)
+        {
+            const char* id = cameraIds->cameraIds[i];
+            auto cameraDevice{ std::make_shared<CameraDevice>()};
+
+            API24::ACameraMetadata *metadataObj;
+            GET_CAMERA_FUNCTION(ACameraManager_getCameraCharacteristics)(m_implData->cameraManager, id, &metadataObj);
+
+            // Get all available stream configurations supported by the camera
+            API24::ACameraMetadata_const_entry streamConfigurations = {};
+            GET_CAMERA_FUNCTION(ACameraMetadata_getConstEntry)(metadataObj, API24::ACAMERA_SCALER_AVAILABLE_STREAM_CONFIGURATIONS, &streamConfigurations);
+            // format of the data:
+            // 0: format
+            // 1: width
+            // 2: height
+            // 3: input?
+
+            for (uint32_t j = 0; j < streamConfigurations.count; j += 4) {
+                int32_t format{streamConfigurations.data.i32[j + 0]};
+                int32_t width{streamConfigurations.data.i32[j + 1]};
+                int32_t height{streamConfigurations.data.i32[j + 2]};
+                int32_t input{streamConfigurations.data.i32[j + 3]};
+
+                if (input || format != AIMAGE_FORMAT_YUV_420_888)
+                {
+                    // Ignore the configuration if it is either an input type or not a preview format
+                    continue;
+                }
+
+                auto cameraResolution{ std::make_shared<CameraTrack>() };
```

Comment:
I don't think you want a shared_ptr here. My intuition think that the supportedResolutions vector should own the struct, so this should be a plain std::vector<CameraTrack>. You would construct the camera track by using supportedResolutions.emplace_back() which would create the object, then pull it off the back and edit the values.

Take a look how something similar is done in AR.
https://github.com/BabylonJS/BabylonNative/blob/581b4b4e16a46e31b555204472886eb3fbf77418/Dependencies/xr/Source/ARCore/XR.cpp#L1113-L1116

Reply:
I updated the CameraTrack class to now properly support the pimpl patter and the move operator so that the vector can contain just the class itself and not a shared or unique pointer to it.